qualcommbe: enable v6.12 testing kernel

mrnuke opened the pull request:
@Ansuel, @robimarko, this is my attempt to bring **qualcommbe** on v6.12 kernel. I am testing this on a TP-Link BE550 (dts not included).

I have one of the ethernet ports working and passing traffic, which proves that the ppe driver works. The other four ports use a QCA8084, whose clocks I was not able to figure out on ether v6.6 or v6.12.
- ppe -> **uniphy2** -> QCA8081 works okay, 2.5G okay
- ppe -> **uniphy0** -> QCA8084 gets stuck on the qca8k clocks

qcom,ppe:
---------

The first half of the driver has a newer submission from Feb 2025, which applies cleanly to v6.12. There is also a 13 May 2025 submission that I did not get a chance to look at yet. The second half of the driver was ported from target/linux/qualcommbe/patches-6.6. Edits done to the patches were noted in the commit messages

qcom,ipq9574-pcs:
-----------------

A newer version is available, which uses a different Kconfig symbol and .compatible. I chose this version (PATCH v5) because it incorporates review feedback, and is likely closer to the final version, should it ever be accepted upstream.

Some of the previous fixes (6.6 patches 700 -> 702) did not port over. I think that it's best to give the original submitter a chance to look over how they might apply.

The patches to add "10gbase-r", "2500base-x", "1000base-x", and "10g-qxgmii" modes have also been ported. I tested "2500base-x".

QCA8084:
--------

Taken directly from target/linux/qualcommbe/patches-6.6, and is the cleanest version I could find.

mrnuke committed:
qualcommbe: enable v6.12 testing kernel

Now that patches for v6.12 are available, add the config, and enable
6.12 as a testing kernel.

Compared to v6.6 the patches to add "10gbase-r", "2500base-x",
"1000base-x", and "10g-qxgmii" modes have not been ported yet. They do
not apply cleanly to the new PCS driver. As the currently supported
hardware does not use those modes, it is not possible to test them at
this time.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18796
Signed-off-by: Robert Marko <robimarko@gmail.com>

mrnuke committed:
qualcommbe: v6.12: add upstreamed patches

Add relevant patches from upstream, up to v5.16-rc6. The gaps in the
patch numbersing are either patches that were picked into the stable
kernel (6.12.y), or that are already backported in
target/linux/generic. The gaps makes it easy for me to pick these
patches from my working kernel git branch.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18796
Signed-off-by: Robert Marko <robimarko@gmail.com>

mrnuke committed:
qualcommbe: v6.12: add pending patches

Add dts fixes from linux-next. Two patches from the NSSCC series are
still in -next did not yet land in mainline, as well as misc other DTS
changes. Add them here.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18796
Signed-off-by: Robert Marko <robimarko@gmail.com>

mrnuke committed:
qualcommbe: v6.12: add various DTS updates for IPQ9574 and RDP433

These devicetree updates can work with the existing upstreamcode. They
do not require code changes. Add them before any code change.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18796
Signed-off-by: Robert Marko <robimarko@gmail.com>

mrnuke committed:
qualcommbe: v6.12: add QCA8084 ethernet PHY driver

This driver is cherry-picked from target/linux/qualcommbe/patches-6.6.
While Qualcomm did submit past patches for QCA8084, the code has since
ben split from at803x. The existing OpenWRT version of the patch is
the cleanest version I could find. Add it here.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18796
Signed-off-by: Robert Marko <robimarko@gmail.com>

mrnuke committed:
qualcommbe: v6.12: add ethernet PCS driver for IPQ9574

Add the v5 of the PCS patch. This is the latest submission as of this
writing. THe last four patches are not part of the submission. They
make the series work with v6.12 kernel, resolve a circular dependency
with the clocks, and add the DTS node. Include them as bundle.

Link: https://lore.kernel.org/lkml/20250207-ipq_pcs_6-14_rc1-v5-0-be2ebec32921@quicinc.com/
Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18796
Signed-off-by: Robert Marko <robimarko@gmail.com>

mrnuke committed:
qualcommbe: v6.12: add pending PPE driver (part 1)

Add v3 submission of the qualcomm PPE driver. As of this writing, it
is the latest version. This lacks the EDMA driver and network device
support. That will be added in part 2.

Link: https://lore.kernel.org/lkml/20250209-qcom_ipq_ppe-v3-0-453ea18d3271@quicinc.com
Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18796
Signed-off-by: Robert Marko <robimarko@gmail.com>

mrnuke committed:
qualcommbe: v6.12: add PPE driver (part 2)

Add the second part of the PPE driver. This includes the EDMA and
network device support. This part does not appear to have been
officially submitted for upstream review. The series is taken from
target/linux/qualcommbe/patches-6.6, and had to be heavily modified
in order to compile of v6.12. Changes to patches are noted in the
respective patch body.

Also add the PPE and EDMA nodes in this series.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18796
Signed-off-by: Robert Marko <robimarko@gmail.com>

mrnuke committed:
qualcommbe: v6.12: enable ethernet for RDP433

Now that ethernet is supported on IPQ9574, add the port nodes and
other requirements to enable ethernet on RDP433.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18796
Signed-off-by: Robert Marko <robimarko@gmail.com>

mrnuke committed:
qualcommbe: v6.12: add NSSCC clock and reset fixes

Port reset and clock fixes from target/linux/qualcommbe/patches-6.6.
Modifications of v6.6 patches are noted in each commit body.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18796
Signed-off-by: Robert Marko <robimarko@gmail.com>

Ansuel commented:
Yep for qca8084 we know it needs some extra love... my dev board had only qca8081. I have only one board with qca8084 but is based on ipq50xx... 

For the PCS patch there is currently war upstream so I doubt we will have an upstream version soon™

robimarko commented:
My IPQ9574 board I had to borrow to somebody from work, I think it has QCA8084.

Will check when I can get it back

mrnuke commented:
My main confusion is regarding how to set up the qca8k input clocks for QCA8084..

What is **pcs_pll0** (chip reference clock), and how to derive it? QCA8084 doc talks about a 50MHz reference clock. Is this the `<&cmn_pll ETH0_50MHZ_CLK>` ?

Are the UNIPHY clocks mislabeled, and mean the clocks for PCS0 and PCS1? Why are there two clocks for uniphy1? How do I provide a 312.5 MHz clock in addition to the uniphy clock (which is probably already at 312.5MHz) ?

The schema describes:
```
        clocks = <&pcs0_pll>,
                 <&qca8k_uniphy0_rx>,
                 <&qca8k_uniphy0_tx>,
                 <&qca8k_uniphy1_rx>,
                 <&qca8k_uniphy1_tx>,
                 <&qca8k_uniphy1_rx312p5m>,
                 <&qca8k_uniphy1_tx312p5m>;

```

And QCA8084 bindings:
```
                 +--| |--+-------------------+--| |--+
                 | PCS1  |<------------+---->| PCS0  |
                 +-------+             |     +-------+
                 |                     |             |
    Ref 50M clk  +--------+            |             |
    ------------>|        | clk & rst  |             |
    GPIO Reset   |QCA8K_CC+------------+             |
    ------------>|        |            |             |
                 +--------+            |             |
                 |                     V             |
                 +--------+--------+--------+--------+
                 |  PHY0  |  PHY1  |  PHY2  |  PHY3  |
                 +--------+--------+--------+--------+
```

mrnuke committed:
qualcommbe: add v6.12 config from v6.6

Copy the v6.6 config to v6.6. It will be updated in a subsequent
commit in order to show the changes.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18796
Signed-off-by: Robert Marko <robimarko@gmail.com>

mrnuke committed:
qualcommbe: v6.12: enable 2.5G and 10G phylink modes in pcs-qcom-ipq9574

The PCS driver in the 6.12 patchset is the v5 submission (see link
below). It solves a number of issues and crashes with teh pcs driver
from the 6.6 patchset. However, this new driver is missing support for
"10gbase-r", "10g-qxgmii", and 1000/2500base-x modes.

Port these modes to the 6.12 patchset. "2500base-x" in particular seems
to be needed to establish a 2.5G link on phy-mode="usxgmii";

Link: https://lore.kernel.org/lkml/20250207-ipq_pcs_6-14_rc1-v5-0-be2ebec32921@quicinc.com/
Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18796
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
@mrnuke I would suggest immediatelly moving to default to 6.12 and dropping 6.6, really no point for it in this target
